lib/portage/tests/util/test_socks5.py

```diff
@@ -269,3 +270,40 @@ def _testSocks5ServerLoopCloseSubprocess():
             shutil.rmtree(tempdir)
 
         return not socks5.proxy.is_running()
+
+
+class Socks5ServerAtExitTestCase(TestCase):
+    """
+    For bug 937384, test that the socks5 proxy is automatically
+    terminated by portage.process.run_exitfuncs(), using a subprocess
+    for isolation.
+    """
+
+    def testSocks5ServerAtExit(self):
+        asyncio.run(self._testSocks5ServerAtExit())
+
+    async def _testSocks5ServerAtExit(self):
+        loop = asyncio.get_running_loop()
+        tempdir = tempfile.mkdtemp()
+        try:
+            pid = await loop.run_in_executor(
+                ForkExecutor(loop=loop), self._testSocks5ServerAtExitSubprocess, tempdir
+            )
+            with self.assertRaises(ProcessLookupError):
+                os.kill(pid, 0)
+        finally:
+            shutil.rmtree(tempdir)
+
+    @staticmethod
+    def _testSocks5ServerAtExitSubprocess(tempdir):
+        loop = global_event_loop()
+        settings = {
+            "PORTAGE_TMPDIR": tempdir,
+            "PORTAGE_BIN_PATH": PORTAGE_BIN_PATH,
+        }
+
+        socks5.get_socks5_proxy(settings)
+        loop.run_until_complete(socks5.proxy.ready())
+        pid = socks5.proxy._proc.pid
+        portage.process.run_exitfuncs()
+        return pid
```

Comment:
Ideally `run_exitfuncs()` would be called automatically by an atexit hook, but this test case currently fails without an explicit `run_exitfuncs()`  call, and it's not just this test case because a standalone script demonstrates a similar problem.

Reply:
The `atexit.register(run_exitfuncs)` call occurs in portage.process module's global scope, but it's ineffective. If I call `atexit.register(portage.process.run_exitfuncs)` at the end of my test script then it works.

Reply:
This issue is somehow related to the `get_socks5_proxy` call, because `atexit.register(portage.process.run_exitfuncs)` is only effective when called after `get_socks5_proxy`.

Reply:
Despite this flaw, the MR is still an improvement over the status quo.

I guess I'll use [this hack](https://stackoverflow.com/questions/16042463/how-can-i-get-the-list-of-registered-atexit-functions-in-python3/63029332#63029332) to examine the registered atexit handlers as a workaround for https://github.com/python/cpython/issues/61388.

Reply:
Apparently it's an issue with an `_exit_function` from the multiprocessing module that triggers this, since `run_exitfuncs` only runs if it's added a second time after the `_exit_function`:

[<function shutdown at 0x7a027b9be5c0>, <function run_exitfuncs at 0x7a027afc5760>, <function _exit_function at 0x7a027a7432e0>]

This explains the interaction with `get_socks5_proxy`.

Reply:
Also, atexit only works in subprocesses after fork since https://github.com/python/cpython/issues/83856 was fixed in python3.13 with no plan for backport.

Reply:
We can use a global variable to trigger `atexit.register(portage.process.run_exitfuncs)` after the first subprocess is started by ForkProcess, in order to prevent the multiprocessing `_exit_function` deadlock.